File: opt/knapsack/solver.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
from copy import copy
from Queue import PriorityQueue
import profile

def solveIt(inputData):
    # Modify this code to run your optimization algorithm

    # parse the input
    lines = inputData.split('\n')

    firstLine = lines[0].split()
    items = int(firstLine[0])
    capacity = int(firstLine[1])

    values = []
    weights = []

    for i in range(1, items+1):
        line = lines[i]
        parts = line.split()

        values.append(int(parts[0]))
        weights.append(int(parts[1]))

    value, taken = depth(capacity, weights, values)

    # prepare the solution in the specified output format
    outputData = str(value) + ' ' + str(1) + '\n'
    outputData += ' '.join(map(str, taken))
    return outputData

def depth(k, weights, values):
    l = len(weights)
    sorted_ratios = ratios = [(values[i] / weights[i], i) for i in range(l)]
    sorted_ratios.sort(reverse=True)

    f = lambda ratios, k, start: estimate(ratios, k, start,  weights, values)

    e = f(sorted_ratios, k, 0)

    s = []

    index = sorted_ratios[0][1]
    s.append((0, 0, k, f(sorted_ratios, k, 1), [0]))
    s.append((0, values[index], k-weights[index], values[index] + f(sorted_ratios, k-weights[index], 1), [1]))
    maxV = 0
    maxToken = []
    while len(s) != 0:
        i, v, w, e, token = s.pop()

        if w >= 0:
            if i == l-1:
                if v > maxV:
                    maxV = v
                    maxToken = copy(token)
            else:
                if e > maxV:
                    index = sorted_ratios[i+1][1]
                    t1 = copy(token)
                    t1.append(0)
                    s.append((i+1, v, w, v + f(sorted_ratios, w, i+1), t1))
                    t2 = copy(token)
                    t2.append(1)
                    s.append((i+1, v+values[index], w-weights[index], v+values[index] + f(sorted_ratios, w-weights[index], i+1), t2))

    token = [0 for i in range(l)]
    for i in range(l):
        token[sorted_ratios[i][1]] = maxToken[i]

    return maxV, token

def best(k, weights, values):
    l = len(weights)
    sorted_ratios = ratios = [(values[i] / weights[i], i) for i in range(l)]
    sorted_ratios.sort(reverse=True)

    f = lambda ratios, k, start: estimate(ratios, k, start,  weights, values)

    e = f(sorted_ratios, k, 0)

    s = PriorityQueue()

    index = sorted_ratios[0][1]
    s.put((-f(sorted_ratios, k, 1), 0, 0, k, [0]))
    s.put((-values[index] - f(sorted_ratios, k-weights[index], 1), 0, values[index], k-weights[index], [1]))
    maxV = 0
    maxToken = []
    while not s.empty():
        e, i, v, w, token = s.get()

        if w >= 0:
            if i == l-1:
                if v > maxV:
                    maxV = v
                    maxToken = copy(token)
            else:
                if e < maxV:
                    index = sorted_ratios[i+1][1]
                    t1 = copy(token)
                    t1.append(0)
                    s.put((v - f(sorted_ratios, w, i+1), i+1, v, w, t1))
                    t2 = copy(token)
                    t2.append(1)
                    s.put((v-values[index] - f(sorted_ratios, w-weights[index], i+1), i+1,  v+values[index], w-weights[index], t2))

    token = [0 for i in range(l)]
    for i in range(l):
        token[sorted_ratios[i][1]] = maxToken[i]

    return maxV, token

def estimate(sorted_ratios, k, start, weights, values):
    l = len(sorted_ratios)
    tempK = k
    e = 0
    for i in range(start, l):
        index = sorted_ratios[i][1]
        if tempK > weights[index]:
            e += values[index]
            tempK -= weights[index]
        else:
            e += values[index] * (tempK / weights[index])
            tempK = 0
    return e


def ps():
    if len(sys.argv) > 1:
        fileLocation = sys.argv[1].strip()
        inputDataFile = open(fileLocation, 'r')
        inputData = ''.join(inputDataFile.readlines())
        inputDataFile.close()
        print solveIt(inputData)
    else:
        print 'This test requires an input file.  Please select one from the data directory. (i.e. python solver.py ./data/ks_4_0)'



import sys

if __name__ == '__main__':
    #profile.run("ps()")
    ps()

File: opt/knapsack/Solver.java
import java.io.*;
import java.lang.ArrayStoreException;
import java.lang.Comparable;
import java.lang.System;
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Stack;

/**
 * The class <code>Solver</code> is an implementation of a greedy algorithm to solve the knapsack problem.
 *
 */
public class Solver {
    
    /**
     * The main class
     */
    public static void main(String[] args) {
        try {
            solve(args);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
    
    /**
     * Read the instance, solve it, and print the solution in the standard output
     */
    public static void solve(String[] args) throws IOException {
        String fileName = null;
        
        // get the temp file name
        for(String arg : args){
            if(arg.startsWith("-file=")){
                fileName = arg.substring(6);
            } 
        }
        if(fileName == null)
            return;
        
        // read the lines out of the file
        List<String> lines = new ArrayList<String>();

        BufferedReader input =  new BufferedReader(new FileReader(fileName));
        try {
            String line = null;
            while (( line = input.readLine()) != null){
                lines.add(line);
            }
        }
        finally {
            input.close();
        }
        
        
        // parse the data in the file
        String[] firstLine = lines.get(0).split("\\s+");
        int items = Integer.parseInt(firstLine[0]);
        int capacity = Integer.parseInt(firstLine[1]);

        int[] values = new int[items];
        int[] weights = new int[items];

        for(int i=1; i < items+1; i++){
          String line = lines.get(i);
          String[] parts = line.split("\\s+");

          values[i-1] = Integer.parseInt(parts[0]);
          weights[i-1] = Integer.parseInt(parts[1]);
        }

        // a trivial greedy algorithm for filling the knapsack
        // it takes items in-order until the knapsack is full
        int value = 0;
        int weight = 0;
        int[] taken = new int[items];

//        for(int i=0; i < items; i++){
//            if(weight + weights[i] <= capacity){
//                taken[i] = 1;
//                value += values[i];
//                weight += weights[i];
//            } else {
//                taken[i] = 0;
//            }
//        }0

        value = depth(capacity, weights, values, taken);


        // prepare the solution in the specified output format
        System.out.println(value+" 1");
        for(int i=0; i < items; i++){
            System.out.print(taken[i]+" ");
        }
        System.out.println("");        
    }

    private static int depth(int k, int[] weights, int[] values, int[] takens){
        int l = weights.length;
        RatioIndex[] ratios = new RatioIndex[l];
        for (int i = 0; i < l; i++){
            double ratio = (double)values[i] / weights[i];
            ratios[i] = new RatioIndex(ratio, i);
        }
        Arrays.sort(ratios);
//        for(int i = 0; i < l; i++){
//            System.out.printf("%f ", ratios[i].Ratio);
//        }
        Stack<State> s = new Stack<State>();

        int index = ratios[0].Index;
        s.push(new State(0, 0, k, estimate(ratios, k, 1, weights, values), 0));
        s.push(new State(0, values[index], k-weights[index], values[index] + estimate(ratios, k-weights[index], 1, weights, values), 1));

        int maxV = 0;

        int[] maxToken = new int[l];

        int end = l - 1;
        while (!s.empty()){
            State state = s.pop();
            takens[state.index] = state.isSelected;
            //if (state.weight >= 0){
                index = state.index;
                if (index == end){
                    if (state.value > maxV){
                        maxV = state.value;
                        maxToken = takens.clone();
                    }
                }
                else{
                    //if (state.estimate > maxV){
                        int newIndex = ratios[index+1].Index;
                        int nextIndex = index + 1;
                        double e = state.value + estimate(ratios, state.weight, nextIndex, weights, values);
                        if (e > maxV){
                            s.push(new State(nextIndex, state.value, state.weight, e, 0));
                        }
                        int w = state.weight - weights[newIndex];
                        if (w > 0)
                        {
                            s.push(new State(nextIndex, state.value + values[newIndex], w,
                                    state.estimate, 1));
                        }
                    //}
                }
            //}
        }

//        for (int i = 0; i < l; i++){
//           System.out.printf("%d ", maxToken[i] );
//        }
        for (int i = 0; i < l; i++){
            takens[ratios[i].Index] = maxToken[i];
        }
        return maxV;
    }

    private static double estimate(RatioIndex[] ratios, int k, int start, int[] weights, int[] values){
        int l = ratios.length;
        int tempK = k;
        double e = 0;

        for (int i = start; i < l; i++){
            int index = ratios[i].Index;
            if (tempK > weights[index]){
                e += values[index];
                tempK -= weights[index];
            }
            else{
                e += values[index] * ((double)tempK / weights[index]);
                break;
            }
        }
        return e;
    }
}